### ISSUE TITLE
New release? -> Testing wanted

### ISSUE THREAD

**Issue by ypid** [opening-hours/opening_hours.js#360]
I know, I have not made a new release for years. My main concern is to be honest that I am not up-to-date with the latest stuff happening in the JavaScript world. And the unreleased work contains a switch in how this library is build and distributed and I have not tested how npm handles this, if users will need to change something and so on.

The question came up in #358:

> How do users of the npm-library get the latest updates?

I assume npm/yarn allow to install from a git repo master branch directly or something like that. If you have tested this or have other feedback about my concern, please let me know.

Cc: @pietervdvn

**Comment by pietervdvn** [comment]
I did a quick try with `npm install git+https://github.com/opening-hours/opening_hours.js.git`, but building my project then didn't work anymore - parcel couldn't resolve the dependency to "opening_hours"

But this was only a quick try though, and my setup probably isn't the cleanest either

**Comment by fodor0205** [comment]
After running `npm install git+https://github.com/opening-hours/opening_hours.js.git`:
The `node_modules/opening_hours/package.json` references ["main": "opening_hours.js"](https://github.com/opening-hours/opening_hours.js/blob/c608a771fec7c7b058de6f562507aeb550d6d18e/package.json#L3), which file is not present in the `node_modules/opening_hours` directory, so `require('opening_hours')` fails.

**Comment by napei** [comment]
I started a fork of this project a while ago to basically rewrite it in modern JS (and even typescript) with a proper build system and tree-shakability but got lost inside its complexity 😂

**Comment by ypid** [comment]
@napei You are my hero. Have you seen #136 which describes my long term plan?

I guess you don’t have your work pushed? https://github.com/napei/opening_hours.js only has my branches. Could be interesting even if you don’t plan to continue on it.

> but got lost inside its complexity

My hope is that with the tests and a working implementation of the complex date math, a rewrite should be doable. The parser is one part I would probably try to write from scratch.

> The inclusion of yaml is also rather annoying, and most of it should be JSON or JS.

That is the only thing I disagree with you. JSON is only for machines. YAML is also for humans and those definitions are meant to be edited by humans (and machines, see my scripts). Sure, we don’t ship the YAML files but bundle them as JS.

**Comment by napei** [comment]
> Have you seen #136 which describes my long term plan?

I think before porting to other languages, it needs to be re-written in modern 2021 Javascript first and split out into small modules that can be worked on incrementally (instead of one `index.js`), even with TypeScript if you want so that it makes life easier later for typechecking.

> My hope is that with the tests and a working implementation of the complex date math, a rewrite should be doable. The parser is one part I would probably try to write from scratch.

If you have the library exported the tests are really simple, just like using the library out in the real world which makes it easy to write tests for. Modern test libraries give you nice code coverage reports and stuff too. But due to the complexity of the strings you can parse there will need to be a lot of tests. Which is also fine.

> JSON is only for machines. YAML is also for humans and those definitions are meant to be edited by humans

Yes that is true. However 99% of people that are working with this library are familiar with JSON because it's identical to POJOs. In that way, it's actually easier to use JSON or even a `export foo = {}` in a `.js` file. Also, it's easier for people who understand the ecosystem to develop with because you get intellisense from JS instead of YAML. But that's of low priority.

I think that's also a good argument for using TypeScript too. You get type safety when you write the code, and you also get a `.d.ts` output that makes every else's life a lot easier when using this library.; gone are the days where JS is just used as a `.min.js` imported into HTML 😂 

> I guess you don’t have pushed your work?

I didn't. But luckily I still have it sitting around locally on a branch that I dedicated to putting things in folders. I'll open a PR if you're interested. (it's not 100% working)

I also have a small nitpick/question because I'm genuinely curious. Where did comments using `{{{` `}}}` come from? Is that some form of sectioning/folding? Now days that's not really a thing anymore and causes clutter, but meh.

**Comment by napei** [comment]
@ypid Update on my branch. Turns out I did a little more than I thought 😂 

I had converted a bunch of stuff to json. So I undid that.

I also added/changed some packages used in the build system to be more in line with what I use usually.
I had also revised it to be faster and more modern(er) so now it generates a bundled version, and a non-bundled version so it's able to be included in a browser without dependencies.

![image](https://user-images.githubusercontent.com/8946502/106063764-d55d4380-614c-11eb-8230-59159e87aa44.png)

**Comment by ypid** [comment]
Sounds good. I also fine with TypeScript.

> Yes that is true. However 99% of people that are working with this library are familiar with JSON because it's identical to POJOs.

I guess lets just leave it like that. I would rather move those definitions out of this project into a common repo, see #300.

> I also have a small nitpick/question because I'm genuinely curious. Where did comments using {{{ }}} come from?

Yes, fold marks, I wrote about it in the README. You are right, it also starts to look to me like clutter.

> I don't really have the time to dedicate to rewriting this library, like you I have other things to do 😄

Too bad, but understandable :) Maybe just open a PR so that it stays visible?

**Comment by napei** [comment]
@ypid Opened #372 also with a list of things that might benefit the project too.

**Comment by ypid** [comment]
Some progress has been made (see the duplicate issue #378 for details). Feel free to test https://www.npmjs.com/package/opening_hours/v/3.6.0-beta.2 :-)

**Comment by KristjanESPERANTO** [comment]
The new version works for me. Still I noticed something strange that I can't explain and what has not occurred before. It concerns the following section in the `index.js`:
```
if (timevar_string[0]) {
    var date_from = SunCalc.getTimes(date, lat, lon)[timevar_string[0]];
    minutes_from  = date_from.getHours() * 60 + date_from.getMinutes() + timevar_add[0];
}
```

After the build process, the section in the `opening_hours+deps.js` looks like this:
```
if (timevar_string[0]) {
    var date_from = suncalc.getTimes(date, lat, lon)[timevar_string[0]];
    date_from.getHours() * 60 + date_from.getMinutes() + timevar_add[0];
}
```
The variable `name minutes_from` is suddenly missing in the third line. So this line no longer has any effect.

This section occurs more frequently in the file. This behavior only occurs at one point.

**Comment by happyfloat** [comment]
@ypid I'm currently using the beta version and it works well so far! Thank you very much for maintaining this library! If my startup has any success, hopefully my team can contribute back.

**Comment by ypid** [comment]
I fixed my website deployment in https://github.com/opening-hours/opening_hours_map/commit/35a39492feb6c69c744f3fa23de2852075147c16 and deployed to https://openingh.ypid.de/

> The new version works for me. Still I noticed something strange that I can't explain and what has not occurred before. It concerns the following section in the index.js:

@KristjanESPERANTO: Well spotted! In this once occurrence, the variable was not used so it was optimized out I assume. I fixed it in https://github.com/opening-hours/opening_hours.js/commit/7dd8a3db2ee5a9ae8240c7a41ae7b82a96647756

So I guess making a release in a few weeks is better than the broken beta release that was published as stable.
